File: src/lib/mix.ts
// The mix: which tracks are audible, and how loud, at a given moment.
//
// Two consumers have to agree exactly — the preview player (PreviewPlayer) and
// the renderer (worker/mix.py, its Python twin). If they disagree, the preview
// is a lie about what the render will sound like, and nothing on screen says
// so. Everything here is pure and unit-tested; the Python side mirrors it
// function for function.
import type { Clip, Track } from "./db/types";

/** One automation point: a lane's output level at a moment in TIMELINE time. */
export interface AutoPoint {
  t_ms: number;
  gain_db: number;
}

/** The drawable/editable range of the automation lane. Wider than a fader
 *  needs to be useful and narrow enough that a 40px lane still resolves ~1dB
 *  per two pixels. */
export const AUTO_MIN_DB = -30;
export const AUTO_MAX_DB = 6;
/** Past this a curve is a drawing, not an edit — and the renderer compiles
 *  every point into one nested ffmpeg expression. */
export const MAX_AUTO_POINTS = 64;

export const clampDb = (db: number) => Math.min(AUTO_MAX_DB, Math.max(AUTO_MIN_DB, db));

/** dB -> linear multiplier, floored at silence. -inf isn't representable in a
 *  slider, so the bottom of the range IS silence. */
export function linearGain(db: number): number {
  if (!Number.isFinite(db) || db <= AUTO_MIN_DB) return 0;
  return Math.pow(10, db / 20);
}

/** Media-element volume: the same curve, clamped to what HTMLMediaElement
 *  accepts (it throws on > 1, and boosting isn't possible in the browser —
 *  the renderer does honour gain above 0dB). */
export const elementVolume = (db: number) => Math.min(1, Math.max(0, linearGain(db)));

export const sortPoints = (pts: AutoPoint[]): AutoPoint[] =>
  [...pts].sort((a, b) => a.t_ms - b.t_ms);

/** Validated+sorted copy per points ARRAY, cached by identity. The preview
 *  player asks for lane levels on every rAF tick for every clip, and
 *  re-filtering and re-sorting an unchanged array 60×/s per clip was real
 *  main-thread work (and garbage) on WKWebView. Edits replace the array —
 *  the store never mutates one in place — so identity is an exact cache key,
 *  and a WeakMap lets dropped curves collect. */
const sortedCache = new WeakMap<AutoPoint[], AutoPoint[]>();
function sortedValidPoints(points: AutoPoint[] | null | undefined): AutoPoint[] {
  if (!points || !points.length) return [];
  let s = sortedCache.get(points);
  if (!s) {
    s = sortPoints(points.filter((p) => Number.isFinite(p?.t_ms) && Number.isFinite(p?.gain_db)));
    sortedCache.set(points, s);
  }
  return s;
}

/** The lane's level at `ms`: piecewise-linear between points, flat before the
 *  first and after the last. No points at all means the curve isn't in play
 *  and the fader value stands. */
export function gainAtMs(points: AutoPoint[] | null | undefined, ms: number, fallbackDb = 0): number {
  const pts = sortedValidPoints(points);
  if (!pts.length) return fallbackDb;
  if (ms <= pts[0].t_ms) return pts[0].gain_db;
  const last = pts[pts.length - 1];
  if (ms >= last.t_ms) return last.gain_db;
  for (let i = 1; i < pts.length; i++) {
    const a = pts[i - 1];
    const b = pts[i];
    if (ms <= b.t_ms) {
      const span = b.t_ms - a.t_ms;
      if (span <= 0) return b.gain_db;
      return a.gain_db + ((b.gain_db - a.gain_db) * (ms - a.t_ms)) / span;
    }
  }
  return last.gain_db;
}

type MixTrack = Pick<Track, "id" | "kind" | "muted" | "solo" | "gain_db"> & {
  automation?: AutoPoint[] | null;
};

/** Is anything soloed anywhere on the timeline? Solo is global by design: it
 *  answers "let me hear just this", so it has to silence the video lanes'
 *  baked audio too, not only the other audio lanes. */
export const soloActive = (tracks: MixTrack[]): boolean => tracks.some((t) => t.solo);

/** Mute wins over the track's own solo — a lane that is both is silent, which
 *  is what every NLE does and the only reading that lets you solo a group and
 *  still drop one member. */
export function isAudible(track: MixTrack | undefined, tracks: MixTrack[]): boolean {
  if (!track) return false;
  if (track.muted) return false;
  return !soloActive(tracks) || !!track.solo;
}

/** The lane's contribution at `ms`, in dB. Automation REPLACES the fader when
 *  it has points (see the migration note) — reading them as an offset would
 *  make every curve depend on where the fader happened to be. */
export function trackGainDbAt(track: MixTrack | undefined, ms: number): number {
  if (!track) return 0;
  const pts = track.automation ?? [];
  return pts.length ? gainAtMs(pts, ms, Number(track.gain_db) || 0) : Number(track.gain_db) || 0;
}

/** What one clip should actually be playing back at, right now: lane level +
 *  the clip's own trim, or silence if its lane isn't in the mix. A video
 *  clip whose audio has been detached is silent on the video lane — its audio
 *  is a separate clip on an audio lane now, and playing both doubles it. */
export function clipVolume(
  clip: Pick<Clip, "gain_db" | "track_id"> & { audio_detached?: boolean | null },
  tracks: MixTrack[],
  ms: number,
  /** Pre-resolved lane, for callers in a per-frame loop — the find below is an
   *  O(tracks) scan per clip per tick otherwise. Same answer either way. */
  track: MixTrack | undefined = undefined
): number {
  track ??= tracks.find((t) => t.id === clip.track_id);
  if (!isAudible(track, tracks)) return 0;
  if (clip.audio_detached && track?.kind === "video") return 0;
  return elementVolume(trackGainDbAt(track, ms) + (Number(clip.gain_db) || 0));
}

// ------------------------------------------------------- automation editing --

/** Add a point, replacing any within `epsMs` of it — clicking twice in the
 *  same spot must adjust that point, not stack an invisible second one on top
 *  of it. */
export function addPoint(points: AutoPoint[], p: AutoPoint, epsMs = 40): AutoPoint[] {
  const t = Math.max(0, Math.round(p.t_ms));
  const kept = (points ?? []).filter((q) => Math.abs(q.t_ms - t) > epsMs);
  if (kept.length >= MAX_AUTO_POINTS) return sortPoints(points ?? []);
  return sortPoints([...kept, { t_ms: t, gain_db: clampDb(p.gain_db) }]);
}

/** Move point `idx`. Time is free (the list re-sorts) so a point dragged past
 *  its neighbour reorders instead of sticking. */
export function movePoint(points: AutoPoint[], idx: number, p: AutoPoint): AutoPoint[] {
  const pts = points ?? [];
  if (idx < 0 || idx >= pts.length) return sortPoints(pts);
  const next = pts.map((q, i) =>
    i === idx ? { t_ms: Math.max(0, Math.round(p.t_ms)), gain_db: clampDb(p.gain_db) } : q
  );
  return sortPoints(next);
}

export function removePoint(points: AutoPoint[], idx: number): AutoPoint[] {
  return sortPoints((points ?? []).filter((_, i) => i !== idx));
}

/** Which point a pointer landed on, in the lane's own pixel space. */
export function hitPoint(
  points: AutoPoint[],
  at: { x: number; y: number },
  toXY: (p: AutoPoint) => { x: number; y: number },
  radiusPx = 7
): number {
  let best = -1;
  let bestD = radiusPx;
  (points ?? []).forEach((p, i) => {
    const { x, y } = toXY(p);
    const d = Math.hypot(x - at.x, y - at.y);
    if (d <= bestD) {
      best = i;
      bestD = d;
    }
  });
  return best;
}

/** dB -> fraction down the automation lane (0 = top). */
export const dbToFrac = (db: number) => (AUTO_MAX_DB - clampDb(db)) / (AUTO_MAX_DB - AUTO_MIN_DB);
/** The inverse, for a pointer landing in the lane. */
export const fracToDb = (frac: number) =>
  clampDb(AUTO_MAX_DB - Math.min(1, Math.max(0, frac)) * (AUTO_MAX_DB - AUTO_MIN_DB));

File: worker/mix.py
"""The mix: which timeline tracks are audible, and how loud, over time.

The Python twin of `src/lib/mix.ts`. The preview player answers these
questions in the browser and this answers them for the render; if the two
disagree the preview is a lie about the finished file, and nothing on screen
says so. Pure functions, tested by worker/tests/test_mix.py against the same
cases as src/lib/mix.test.ts.

Automation is a list of ``{"t_ms": int, "gain_db": float}`` in TIMELINE time,
piecewise-linear between points and flat outside them, and it REPLACES the
lane fader when it has any points.
"""

import audio_fx

AUTO_MIN_DB = -30.0
AUTO_MAX_DB = 6.0


def _num(v, default=0.0):
    try:
        f = float(v)
    except (TypeError, ValueError):
        return default
    return f if f == f and abs(f) != float("inf") else default   # NaN/inf out


def clamp_db(db):
    return max(AUTO_MIN_DB, min(AUTO_MAX_DB, _num(db)))


def linear_gain(db):
    """dB -> linear multiplier. The bottom of the range IS silence: -inf has
    no slider position, so the browser and ffmpeg both treat it as 0."""
    d = _num(db)
    if d <= AUTO_MIN_DB:
        return 0.0
    return 10.0 ** (d / 20.0)


def sort_points(points):
    pts = []
    for p in (points or []):
        if not isinstance(p, dict):
            continue
        if "t_ms" not in p or "gain_db" not in p:
            continue
        pts.append({"t_ms": _num(p.get("t_ms")), "gain_db": _num(p.get("gain_db"))})
    return sorted(pts, key=lambda p: p["t_ms"])


def gain_at_ms(points, ms, fallback_db=0.0):
    pts = sort_points(points)
    if not pts:
        return _num(fallback_db)
    if ms <= pts[0]["t_ms"]:
        return pts[0]["gain_db"]
    if ms >= pts[-1]["t_ms"]:
        return pts[-1]["gain_db"]
    for i in range(1, len(pts)):
        a, b = pts[i - 1], pts[i]
        if ms <= b["t_ms"]:
            span = b["t_ms"] - a["t_ms"]
            if span <= 0:
                return b["gain_db"]
            return a["gain_db"] + (b["gain_db"] - a["gain_db"]) * (ms - a["t_ms"]) / span
    return pts[-1]["gain_db"]


def solo_active(tracks):
    """Solo is global: it answers "let me hear just this", so it has to reach
    the video lanes' baked audio as well as the other audio lanes."""
    return any(t.get("solo") for t in (tracks or []))


def is_audible(track, tracks):
    """Mute wins over the lane's own solo — the only reading that lets you
    solo a group and still drop one member."""
    if not track:
        return False
    if track.get("muted"):
        return False
    return (not solo_active(tracks)) or bool(track.get("solo"))


def audible_tracks(tracks, kind=None):
    return [t for t in (tracks or [])
            if (kind is None or t.get("kind") == kind) and is_audible(t, tracks)]


def track_points(track):
    return sort_points((track or {}).get("automation") or [])


def track_gain_db_at(track, ms):
    pts = track_points(track)
    base = _num((track or {}).get("gain_db"))
    return gain_at_ms(pts, ms, base) if pts else base


def volume_filter(track, clip_gain_db=0.0, *, offset_ms=0):
    """The ffmpeg `volume` filter for one audio clip on `track`.

    Applied AFTER `adelay`, so the filter's `t` is timeline seconds and the
    automation's own timebase needs no shifting. Without automation this is
    the constant the renderer always used; with it, one nested expression
    evaluated per frame.

    `offset_ms` shifts the curve for a piece of media that is NOT at timeline
    zero after delay (unused today — adelay puts every clip on the timeline
    clock — but the parameter is what makes that assumption explicit).
    """
    pts = track_points(track)
    gain = _num(clip_gain_db)
    if not pts:
        return "volume=%.4fdB" % (_num((track or {}).get("gain_db")) + gain)
    expr = _volume_expr(pts, gain, offset_ms=offset_ms)
    return "volume=volume='%s':eval=frame" % expr


def lane_bus(track, labels, out_label):
    """-> the filtergraph fragment that sums one lane through its own rack, or
    None when the lane has no rack and its clips go straight to the mix.

    Render-only, like `volume_filter` above — the preview builds the same stage
    as Web Audio nodes (`src/lib/audioGraph.ts`), where a bus is a GainNode and
    needs no string. What both halves must agree on is the ORDER, which is the
    reason this is one function rather than four lines inside the renderer:

        clips (each already carrying its own inserts and its own gain)
          -> sum -> pad -> the LANE's inserts -> the lane fader/automation

    Two of those are easy to get backwards and silent when wrong:

      - The fader is LAST. Console order, and the only order in which a lane
        compressor hears the lane instead of the fader. It is free to move:
        gain is linear and distributes over a sum, so this computes exactly
        what the per-clip form computed, which is why a lane with no rack can
        keep that form untouched.
      - The PAD comes before the inserts. A reverb or an echo goes on making
        sound after its input stops and an ffmpeg stream ends where its last
        clip does, so without it the tail is cut at the last clip. The final
        mix takes its length from the video and trims the pad back off.
    """
    fx = audio_fx.filters((track or {}).get("audio_fx"))
    if not fx or not labels:
        return None
    steps = []
    if len(labels) > 1:
        # `longest`, not `first`: a lane's later clips must not be truncated by
        # whichever one happens to end soonest.
        steps.append("amix=inputs=%d:normalize=0:duration=longest" % len(labels))
    tail = audio_fx.tail_ms((track or {}).get("audio_fx"))
    if tail > 0:
        steps.append("apad=pad_dur=%.3f" % (tail / 1000.0))
    steps.extend(fx)
    steps.append(volume_filter(track))
    return "%s%s%s" % ("".join(labels), ",".join(steps), out_label)


def _lin(db):
    return linear_gain(db)


def _volume_expr(points, clip_gain_db=0.0, *, offset_ms=0):
    """Piecewise-linear automation as one ffmpeg expression in LINEAR gain.

    Interpolating in dB and converting per segment endpoint would be more
    correct to the ear, but ffmpeg's expression language has no log10 on the
    literal we need it for; interpolating the linear values between converted
    endpoints matches what the browser's <audio>.volume does frame to frame,
    which is the parity that actually matters here.
    """
    pts = sort_points(points)
    lin = [(max(0.0, (p["t_ms"] - offset_ms) / 1000.0), _lin(p["gain_db"] + clip_gain_db))
           for p in pts]
    if len(lin) == 1:
        return "%.6f" % lin[0][1]
    expr = "%.6f" % lin[-1][1]          # after the last point: hold
    for i in range(len(lin) - 1, 0, -1):
        t0, g0 = lin[i - 1]
        t1, g1 = lin[i]
        span = t1 - t0
        if span <= 0:
            seg = "%.6f" % g1
        elif abs(g1 - g0) < 1e-9:
            seg = "%.6f" % g0
        else:
            seg = "(%.6f+(%.6f)*(t-%.6f))" % (g0, (g1 - g0) / span, t0)
        expr = "if(lt(t,%.6f),%s,%s)" % (t1, seg, expr)
    # before the first point: hold the first value
    return "if(lt(t,%.6f),%.6f,%s)" % (lin[0][0], lin[0][1], expr)


def eval_volume_expr(expr, t):
    """Evaluate what `_volume_expr` produced, for tests. Not used at render
    time — ffmpeg is the evaluator there — but a filter string nobody can read
    back is a filter string nobody can check."""
    expr = expr.strip()
    if not expr.startswith("if("):
        return float(expr)
    # if(lt(t,T),A,B) — split on the top-level commas of the outer call
    inner = expr[3:-1]
    parts, depth, cur = [], 0, ""
    for ch in inner:
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        if ch == "," and depth == 0:
            parts.append(cur)
            cur = ""
        else:
            cur += ch
    parts.append(cur)
    cond, a, b = parts
    thresh = float(cond[cond.index(",") + 1:-1])
    if t < thresh:
        return _eval_seg(a, t)
    return eval_volume_expr(b, t)


def _eval_seg(seg, t):
    seg = seg.strip()
    if seg.startswith("if("):
        return eval_volume_expr(seg, t)
    if not seg.startswith("("):
        return float(seg)
    body = seg[1:-1]                       # g0+(slope)*(t-t0)
    g0, rest = body.split("+(", 1)
    slope, rest = rest.split(")*(t-", 1)
    t0 = rest[:-1]
    return float(g0) + float(slope) * (t - float(t0))
